tests/worker/test_workflow.py

```diff
@@ -3213,21 +3216,57 @@ async def test_manual_result_type(client: Client):
         assert res4 == ManualResultType(some_string="from-query")
 
 
+@activity.defn
+async def wait_forever_activity() -> None:
+    await asyncio.Future()
+
+
+@workflow.defn
+class WaitForeverWorkflow:
+    @workflow.run
+    async def run(self) -> None:
+        await asyncio.Future()
+
+
 @workflow.defn
-class SwallowGeneratorExitWorkflow:
+class CacheEvictionTearDownWorkflow:
     def __init__(self) -> None:
         self._signal_count = 0
 
     @workflow.run
     async def run(self) -> None:
+        # Start several things in background. This is just to show that eviction
+        # can work even with these things running.
+        tasks = [
+            asyncio.create_task(
+                workflow.execute_activity(
+                    wait_forever_activity, start_to_close_timeout=timedelta(hours=1)
+                )
+            ),
+            asyncio.create_task(
+                workflow.execute_child_workflow(WaitForeverWorkflow.run)
+            ),
+            asyncio.create_task(asyncio.sleep(1000)),
+            asyncio.shield(
+                workflow.execute_activity(
+                    wait_forever_activity, start_to_close_timeout=timedelta(hours=1)
+                )
+            ),
+        ]
+        gather_fut = asyncio.gather(*tasks, return_exceptions=True)
         try:
             # Wait for signal count to reach 2
+            await asyncio.sleep(0.01)
             await workflow.wait_condition(lambda: self._signal_count > 1)
         finally:
-            # This finally, on eviction, is actually called because the above
-            # await raises GeneratorExit. Then this will raise a
-            # _NotInWorkflowEventLoopError swallowing that.
+            # This finally, on eviction, is actually called but the command
+            # should be ignored
+            await asyncio.sleep(0.01)
             await workflow.wait_condition(lambda: self._signal_count > 2)
+            # Cancel gather tasks and wait on them, but ignore the errors
+            for task in tasks:
+                task.cancel()
+            await gather_fut
```

Comment:
So if you spawned a task and *didn't* wait for it here, that'll cause the deadlock and prevent shutdown, right?

Will that task see a cancel or the generator exit raised within it? If so, and it exits, then shutdown will proceed, correct? 

Might be worth adding a task like that that isn't properly gathered here, but still can exit w/o preventing shutdown.

Reply:
> So if you spawned a task and didn't wait for it here, that'll cause the deadlock and prevent shutdown, right?

No, you are not required to wait for task completion (Python gives you a warning if you don't sometimes in general but we don't). Deadlock only happens if you do something bad sandbox couldn't catch. Using regular asyncio is always interruptible.

> Will that task see a cancel or the generator exit raised within it? If so, and it exits, then shutdown will proceed, correct?

Upon eviction, we cancel outstanding tasks, so they're all a cancel, but we ignore side effects.

> Might be worth adding a task like that that isn't properly gathered here, but still can exit w/o preventing shutdown.

Will do

Reply:
Added more background tasks